File: lessons/02-variables-types/VariablesTypes.java
// VariablesTypes.java
public class VariablesTypes {
    public static void main(String[] args) {
        // Primitive types
        byte smallNum = 127;           // 8-bit
        short mediumNum = 32000;       // 16-bit
        int age = 25;                  // 32-bit
        long bigNum = 9223372036854775807L;  // 64-bit

        float price = 19.99f;          // 32-bit float
        double precise = 3.14159265359; // 64-bit float

        char letter = 'A';             // Unicode character
        boolean isStudent = true;

        // Reference types (objects)
        String name = "Alice";         // String is an object
        Integer boxedAge = 25;         // Wrapper for int

        // Type inference (Java 10+)
        var inferredType = "I'm a String";  // type inferred
        var number = 42;               // inferred as int

        // Cannot change type
        // age = "text";  // ERROR: incompatible types

        System.out.println("Age: " + age);
        System.out.println("Name: " + name);
        System.out.println("Price: " + price);
        System.out.println("Letter: " + letter);
        System.out.println("Is student: " + isStudent);
    }
}

File: lessons/02-variables-types/variables_types.rs
// variables_types.rs

fn main() {
    // Immutable by default
    let age = 25;               // inferred as i32
    let price = 19.99;          // inferred as f64
    let name = "Alice";         // &str (string slice)
    let is_student = true;      // bool

    // Explicit type annotations
    let explicit_age: i32 = 25;
    let explicit_price: f64 = 19.99;

    // Mutable variables (must be declared)
    let mut counter = 0;
    counter = counter + 1;      // OK
    // age = 30;                // ERROR: cannot assign twice

    // Integer types (signed and unsigned)
    let signed: i32 = -42;      // i8, i16, i32, i64, i128
    let unsigned: u32 = 42;     // u8, u16, u32, u64, u128

    // Floating point
    let f32_num: f32 = 3.14;
    let f64_num: f64 = 3.14159265359;

    // Character (Unicode)
    let letter: char = 'A';
    let emoji: char = '😀';

    // Tuples
    let person: (&str, i32) = ("Alice", 25);

    // Arrays (fixed size)
    let numbers: [i32; 5] = [1, 2, 3, 4, 5];

    println!("Age: {}", age);
    println!("Name: {}", name);
    println!("Is student: {}", is_student);
    println!("Person: {:?}", person);
    println!("Numbers: {:?}", numbers);
    println!("Letter: {}, Emoji: {}", letter, emoji);
}
